---
title: "HaploReconstruct"
author: "Changyi, Neda"
date: '2025'
output: html_document
---

Load needed packages to use haploreconstruct & haplovalidate
```{r setup, include=FALSE}
library(data.table)
options(scipen=999)
library(haplovalidate)
library(haploReconstruct)
library(randomcoloR)
library(ggpubr)
```

## HAPLORECONSTRUCT

load functions (+ required packages)
```{r}

source("02functions/EEC.freq.traj.R")
source("02functions/EEC.manhattan.plot.R")
sessionInfo()
```


function for clustering

```{r}
perform_clustering<-function(l.freqs, l.min.cl.cor=min.cl.cor, l.min.minor.freq=min.minor.freq, l.max.minor.freq=max.minor.freq,
                             l.min.freq.change=min.freq.change,l.min.repl=min.repl,l.min.cl.size=min.cl.size,
                             l.win.size=win.size,l.ngen=ngen,l.run.id="01",l.chr=c("2","3")){
  print("Initialize Time Series ...")
  l.markers<-data.frame("chr"=character(),"pos"=numeric(),"cluster"=numeric())
  #perform haplotype reconstruct for the chromosomes present in your data:
  for(l.chr.iter in l.chr){
  #format time series data 
  temp.freqs<-subset(l.freqs,chr==l.chr.iter)
  timeSeries<-initialize_SNP_time_series(chr=temp.freqs$chr,pos=temp.freqs$pos,base.freq=temp.freqs$basePops,
                                         lib.freqs=temp.freqs[,grep("L",colnames(temp.freqs)),with=F],pop.ident=rep(c(1:10),l.ngen),
                                         pop.generation = rep(seq(0,l.ngen*10-10,10),each=10),
                                         use.libs = rep(T,10*l.ngen), min.minor.freq=l.min.minor.freq,
                                         max.minor.freq = l.max.minor.freq, winsize = l.win.size, min.lib.frac = 0.75,
                                         minfreqchange = l.min.freq.change, win.scale = "bp", minrepl = l.min.repl)
 
  print(paste("Clustering on chromosome",l.chr.iter,"is running. Please be patient ...",sep=" "))
  hbs<-reconstruct_hb(timeSeries,chrom=l.chr.iter, min.cl.size=l.min.cl.size, min.cl.cor=l.min.cl.cor,min.inter=4,single.win=T)
  
  if (number_hbr(hbs)!=0){
      n.clusters<-number_hbr(hbs)
      for (j in c(1:n.clusters)){
        temp.markers<-data.frame("chr"=rep(l.chr.iter,length(markers(hbs,j))),"pos"=markers(hbs,j),"cluster"=rep(j,length(markers(hbs,j))))
        l.markers<-rbind(l.markers,temp.markers)
      }
  }
  }
  #arrange & format result 
  l.markers<-arrange(l.markers,chr,pos)
  l.markers$id<-paste(l.markers$chr,l.markers$cluster,sep=".")
  #save clustering as .rds 
  toSave<-paste("HaploReconstruct-Min-Cl-Corr",l.min.cl.cor,"-Min-Minor-Freq",l.min.minor.freq,
                "-Max-Minor-Freq",l.max.minor.freq,"-Min-Freq-Change",l.min.freq.change,"-Min-Repl",l.min.repl,
                "-Min-Cl-Size",l.min.cl.size,"-Win-Size",l.win.size,"Run-ID",l.run.id,".rds",sep="")
  saveRDS(hbs,file = toSave)
  return(l.markers)
}

obtain_colors<-function(x,y){
  n <- length(unique(x$id))
  palette <- distinctColorPalette(n)
  x$id<-factor(x$id)
  x$color<-x$id
  levels(x$color)<-palette
  z<-merge(x,y,by=c("chr","pos"))
  return(z)
}
```



data
```{r}
#target sites:
targets<-readRDS("02data/929_targets_all.rds")
#ACER CMH results:
cmh<-readRDS("02data/sim929_acer_cmh.rds")
colnames(cmh)<-c("chr","pos","value")
#sync file:
candSNP<-readRDS("02data/sim929_cands.rds")
```


Explore the data
```{r}
head(cmh)
#View(cmh)
head(candSNP)
#View(cands)
```
cmh is storing a table with 1) chromosome 2) position 3)log10(pval), from CMH. This includes signficant and non-significant snps

candSNP includes 1) chromosome 2) position 3) ref allele 4) rising allele 5) falling allele 6) mean starting/base frequency 7-76) frequencies of each replicate across each time point (sorted by time points).(e.g. F0.R1, F0.R2,...F10.R1,F10.R2...F70.R10) To save some memory, cands is prefiltered, but usually this should have the same number of rows as cmh.



## Clustering 

### Parameters for clustering  

```{r}
min.minor.freq<-0 
max.minor.freq<-1
min.freq.change<-0.15
min.repl<-2 #at least two replicates support the clustring(fulfill the min cor)
min.cl.size<-20  #minimum SNP density within the window
win.size<-3e+06  #window size in bp
ngen<-7
min.cl.cor<-0.6

```



actual clustering

```{r}
stringent.cluster<-perform_clustering(l.freqs = candSNP,l.run.id = "001",l.min.cl.cor = min.cl.cor)
table(stringent.cluster$id)

relaxed.cluster<-perform_clustering(l.freqs = candSNP,l.run.id = "002",l.min.cl.cor = 0.2)
table(relaxed.cluster$id)
```


plotting:

```{r}

stringent.cluster<-obtain_colors(stringent.cluster,cmh)
#png("res/demo-stringent.cluster.png",width=750,height=500)
print(get.plot.highlight(data = cmh,l.main="Stringent Clustering",highlight = stringent.cluster,highlight.col = stringent.cluster$color))
#dev.off()

#allele frequency trajectories:
#png("res/demo-stringent.cluster.af.png",width=750,height=500)
print(get.af.traj(l.sync=candSNP,l.cluster=stringent.cluster,l.cluster.id = "2.1"))
print(get.af.traj(l.sync=candSNP,l.cluster=stringent.cluster,l.cluster.id = "3.1"))
#dev.off()

relaxed.cluster<-obtain_colors(relaxed.cluster,cmh)
#png("res/demo-relaxed.cluster.png",width = 750,height = 500)
print(get.plot.highlight(data=cmh,l.main = "Relaxed Clustering",highlight = relaxed.cluster,highlight.col = relaxed.cluster$color))
#dev.off()

#png("res/demo-relaxed.cluster.af.png",width = 750,height = 500)
print(get.af.traj(l.sync = candSNP,l.cluster = relaxed.cluster,l.cluster.id = "2.1"))
print(get.af.traj(l.sync = candSNP,l.cluster = relaxed.cluster,l.cluster.id = "3.1"))
#dev.off()
```

## HAPLORECONSTRUCT SOLUTIONS 

#### Task 0 

```{r}
table(targets$chr)
#7 targets on chr 2, 8 on chr 3 
ggscatter(data=targets,x="freq",y="s")
#negative, non-linear relationship betweenn selection coefficient and starting allele frequency 
```

#### Task 1

```{r}
#number of blocks: 
length(table(stringent.cluster$id))
length(table(relaxed.cluster$id))

obtain_length<-function(x){
  range_temp<-c()
  for(id.iter in unique(x$id)){
    temp<-subset(x,id==id.iter)
    min_temp<-min(temp$pos)
    max_temp<-max(temp$pos)
    range_temp<-c(range_temp,max_temp-min_temp)
  }
  return(range_temp)
}

#SNP number ~ min.cl.cor
marker_SNPs<-data.frame(n=c(table(stringent.cluster$id),table(relaxed.cluster$id)))
marker_SNPs$type<-c(rep("stringent",length(unique(stringent.cluster$id))),rep("relaxed",length(unique(relaxed.cluster$id))))
ggboxplot(data=marker_SNPs,x="type",y="n")

#block length ~ min.cl.cor
block_length<-data.frame(length=c(obtain_length(stringent.cluster),obtain_length(relaxed.cluster)))
block_length$type<-c(rep("stringent",length(unique(stringent.cluster$id))),rep("relaxed",length(unique(relaxed.cluster$id))))
ggboxplot(data=block_length,x="type",y="length")
```

#### Task 2
```{r}
table(targets$pos%in%stringent.cluster$pos)
table(targets$pos%in%relaxed.cluster$pos)

id_stringent<-which(targets$pos%in%stringent.cluster$pos)
pos_stringent<-targets$pos[id_stringent]
table(stringent.cluster$id[stringent.cluster$pos%in%pos_stringent])
#stringent: blocks without targets 

id_relaxed<-which(targets$pos%in%relaxed.cluster$pos)
pos_relaxed<-targets$pos[id_relaxed]
table(relaxed.cluster$id[relaxed.cluster$pos%in%pos_relaxed])
#relaxed: blocks with multiple targets

targets$detected<-rep("No", length(targets$detected))
targets$detected[id_stringent]<-"Yes"
targets$detected<-as.factor(targets$detected)
ggscatter(targets,x="freq",y="s", color = "detected")
#higher s with same freq tend to be caught
```


#### Task 3 

```{r}
relaxed.cluster.small<-perform_clustering(candSNP,l.min.cl.cor = 0.2,l.win.size = 5e+05)
relaxed.cluster.big<-perform_clustering(candSNP,l.min.cl.cor = 0.2, l.win.size = 1e+07)

r<-obtain_colors(relaxed.cluster.small,cmh)
#png("res/demo-relaxed.cluster.s.png",width = 750,height = 500)
print(get.plot.highlight(data=cmh,l.main = "Relaxed Clustering small",highlight = r,highlight.col = r$color))
#dev.off()

#png("res/demo-relaxed.cluster.s.af.png",width = 750,height = 500)
print(get.af.traj(l.sync = candSNP,l.cluster = r,l.cluster.id = "2.1"))
#dev.off()

r<-obtain_colors(relaxed.cluster.big,cmh)
#png("res/demo-relaxed.cluster.b.png",width = 750,height = 500)
print(get.plot.highlight(data=cmh,l.main = "Relaxed Clustering big",highlight = r,highlight.col = r$color))
#dev.off()

#png("res/demo-relaxed.cluster.b.af.png",width = 750,height = 500)
print(get.af.traj(l.sync = candSNP,l.cluster = r,l.cluster.id = "2.1"))
#dev.off()


length(table(relaxed.cluster.big$id))
length(table(relaxed.cluster$id))
length(table(relaxed.cluster.small$id))
#more clusters with decreasing window size 

stringent.cluster.small<-perform_clustering(candSNP,l.min.cl.cor = 0.6,l.win.size = 5e+05)
stringent.cluster.big<-perform_clustering(candSNP,l.min.cl.cor = 0.6,l.win.size = 1e+07)

r<-obtain_colors(stringent.cluster.big,cmh)

#png("res/demo-stringent.cluster.b.af.png",width = 750,height = 500)
print(get.af.traj(l.sync = candSNP,l.cluster = r,l.cluster.id = "2.1"))
#dev.off()

#png("res/demo-stringent.cluster.b.png",width=750,height=500)
print(get.plot.highlight(data = cmh,l.main="Stringent Clustering big",highlight = r,highlight.col = r$color))
#dev.off()

r<-obtain_colors(stringent.cluster.small,cmh)


#png("res/demo-stringent.cluster.s.af.png",width = 750,height = 500)
print(get.af.traj(l.sync = candSNP,l.cluster = r,l.cluster.id = "2.1"))
#dev.off()

#png("res/demo-stringent.cluster.s.png",width=750,height=500)
print(get.plot.highlight(data = cmh,l.main="Stringent Clustering small",highlight = r,highlight.col = r$color))
#dev.off()

length(table(stringent.cluster.big$id))
length(table(stringent.cluster$id))
length(table(stringent.cluster.small$id))

```




## Charaterization of clustered haplotypes
### load haplotype information
```{r}
input_gen=seq(0,60,10)
num_repl=10
f_allinfo=fread('haplotype_snp.sync')



#obtain haplotype ID information
haplo_id=f_allinfo[,c(1,2,86)];colnames(haplo_id)=c("chrom","pos","id")
haplo_id[, chrom_pos := paste(chrom,pos,sep=".")]
haplo_id[, haplo_id := paste(chrom,id,sep=".")]
haplo_id=haplo_id[,c("chrom_pos","haplo_id")]

#merge the haplotype blocks (singlr haplotype encompass the centermere)
haplo_id[, haplo_id := gsub("^2L\\.16", "2L.2R.16", haplo_id)]
haplo_id[, haplo_id := gsub("^2R\\.16", "2L.2R.16", haplo_id)]
haplo_id[, haplo_id := gsub("^3L\\.30", "3L.3R.30", haplo_id)]
haplo_id[, haplo_id := gsub("^3R\\.30", "3L.3R.30", haplo_id)]
```

### extract allele frequency
```{r}
#save the sync file part to extract the allele frequency
sync_f=f_allinfo[,1:73]
fwrite(sync_f,"haplotype_snp_only.sync", col.names=FALSE, sep="\t")

sync_f=read.sync("haplotype_snp_only.sync", gen=rep(input_gen, each=num_repl), repl=rep(1:num_repl,length(input_gen)), polarization = "rising")
#extract allele frequecny
af_mat=poolSeq::af(sync_f, gen = input_gen,repl = 1:num_repl)
loci=rownames(af_mat)

#integrate haplotype ID
af_dt=as.data.table(af_mat, keep.rownames = "chrom_pos")
dt_merged=merge(haplo_id, af_dt, by="chrom_pos", all=TRUE)
cols_for_median=paste0("F",rep(input_gen,num_repl),".R",rep(1:num_repl,each=length(input_gen)),".freq")
dt_median <- dt_merged[, lapply(.SD, median, na.rm = TRUE), by = haplo_id, .SDcols = cols_for_median]

#count num_snps information per block
snp_count=dt_merged[, .N, by=haplo_id]

```

### estimate selection coefficient for each haplotype blocks
```{r}
af=as.matrix(dt_median[,2:ncol(dt_median)])
rownames(af)=dt_median[,haplo_id]

ne<-c(381,335,247,255,246,307,297,244,310,287)
s_table<-data.table(haplotype_id=rownames(af),num_snps=snp_count[,N])
for (repl in 1:num_repl) {
  F0 = paste("F0.R", repl, ".freq", sep='')
  print(F0)
  F60 = paste("F60.R", repl, ".freq", sep='')
  F0=subset(af, select = F0)
  F60=subset(af, select = F60)
  s <- sapply(1:nrow(af), function(x) estimateSH(c(F0[x], F60[x]), Ne=median(ne), t=c(0, 60), h=0.5)$s)
  print(head(s))
  
  s_table[,paste("s", repl, sep="_") :=s]
}
s_table[, median_s := apply(.SD, 1, median, na.rm =TRUE), .SDcols = 3:ncol(s_table)]
fwrite(s_table, file = 'selection_cofficient.txt', sep = "\t", na="NA")
getwd()

```




```{r}
s_info=s_table[,c("haplotype_id", "num_snps", "median_s")]
s_info[, sc_persnp := median_s/num_snps]
colnames(haplo_id)=c("chrom_pos","haplotype_id")
dt1=merge(haplo_id, s_info[, .(haplotype_id, sc_persnp)], by = "haplotype_id", all.x=TRUE)
fwrite(dt1, file = 'snp_sc.txt', sep = "\t", na="NA", quote = FALSE)
```